### DISCUSSION TITLE
How to Query the Shadow Dom Element inside a custom element in a react based application for unit testing purposes?

### DISCUSSION THREAD

**Discussion by shadab-conga** [webdriverio/webdriverio#13966] [Q&A]
I have the following code structure 
<img width="788" alt="Screenshot 2024-12-11 at 1 21 32 PM" src="https://github.com/user-attachments/assets/e5116eaf-7782-400a-87d5-ffe4ccdbaafc">

How to I query the native button element inside the shadow root.

I have tried below methods but they are not seem to be working. Am I doing something incorrect?

<img width="788" alt="Screenshot 2024-12-11 at 1 25 59 PM" src="https://github.com/user-attachments/assets/35ead866-f194-4d86-8393-4e995d735d0b">

The console.log is printing null.

Also the custom element is not interact-able.

I see this point in the docs, but still it does not work by removing the `>>>`.

<img width="993" alt="Screenshot 2024-12-11 at 1 28 20 PM" src="https://github.com/user-attachments/assets/e47973ad-5ea1-442b-8c14-dc0dc33f6417">

**Comment by christian-bromann** [comment]
> How to I query the native button element inside the shadow root.

try `$('button[name="click me"]')`

**Reply by shadab-conga** [reply]
@christian-bromann it still does not work, 

This is my JSX in the react component.

<img width="979" alt="Screenshot 2024-12-11 at 11 21 16 PM" src="https://github.com/user-attachments/assets/55f1a807-c2cc-4b55-bf98-af73b17d5e15" />


I am getting the following error.

<img width="1036" alt="Screenshot 2024-12-11 at 11 22 19 PM" src="https://github.com/user-attachments/assets/2c39f618-7c0a-44a1-90f6-f3396c388ba5" />

**Reply by christian-bromann** [reply]
Do you have a minimal reproduction case?

**Reply by shadab-conga** [reply]
@christian-bromann I am not sure on that, as our component library is in house and not publicly available on npm. 
I will discuss this with my architect and will get back to you.

But if any suggestion you can provide from top of your mind, it will be helpful.

These test cases are running fine just FYI.

<img width="1036" alt="Screenshot 2024-12-11 at 11 48 51 PM" src="https://github.com/user-attachments/assets/27d0e33d-141f-4329-ad71-0f93367d657f" />

Attaching some config options ss.

<img width="1036" alt="Screenshot 2024-12-11 at 11 52 30 PM" src="https://github.com/user-attachments/assets/c76145d5-62ad-4293-81d3-7d4a6f0223b9" />

**Reply by christian-bromann** [reply]
@shadab-conga by default WebdriverIO v9 is able to pierce into the Shadow DOM. There are a lot of variables that may cause why this is not happening for you. Are you using latest WebdriverIO version? A reproduction case would make it easier for us to investigate this issue.

**Reply by shadab-conga** [reply]
@christian-bromann Yes I am using the latest version. Will circle back if I have a reproduction case.
